File: api-kt/src/main/kotlin/site/daydream/mote/service/DriveZipService.kt
package site.daydream.mote.service

import org.jooq.DSLContext
import org.springframework.stereotype.Service
import site.daydream.mote.exception.BadRequestException
import site.daydream.mote.exception.NotFoundException
import site.daydream.mote.model.DriveNode
import java.io.File
import java.io.OutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

@Service
class DriveZipService(
    private val dsl: DSLContext,
    private val driveService: DriveService,
) {
    fun zipFolder(folderId: Long, out: OutputStream) {
        val root = driveService.findById(folderId)
        if (root.type != "folder" || root.deletedAt != null) {
            throw BadRequestException("drive node not found")
        }
        val descendants = driveService.collectDescendants(folderId)

        ZipOutputStream(out).use { zw ->
            val seen = HashSet<String>()
            for (d in descendants) {
                if (d.id == folderId) continue
                var rel = d.relPath.removePrefix(root.name).removePrefix("/")
                rel = sanitizeZipPath(rel)
                if (rel.isBlank()) continue
                val name = if (d.type == "folder") "$rel/" else rel
                if (!seen.add(name)) continue
                zw.putNextEntry(ZipEntry(name))
                if (d.type == "file" && !d.blobPath.isNullOrBlank()) {
                    copyBlobInto(zw, d.blobPath)
                }
                zw.closeEntry()
            }
        }
    }

    /**
     * Resolves a multi-select into zip targets: dedups ids, drops ids nested
     * under other selected ids (their content arrives via the ancestor) and
     * skips deleted/missing nodes. Throws NotFoundException when nothing
     * remains — callers resolve BEFORE streaming so a clean 404 is still
     * possible.
     */
    fun zipTargets(ids: List<Long>): List<DriveNode> {
        val uniq = ids.distinct()
        val nested = nestedSelections(uniq)
        val targets = uniq.asSequence()
            .filter { it !in nested }
            .mapNotNull { driveService.findByIdOrNull(it) }
            .filter { it.deletedAt == null }
            .toList()
        if (targets.isEmpty()) throw NotFoundException("drive node not found")
        return targets
    }

    /**
     * Streams a zip archive of the given nodes. Unlike zipFolder — which
     * strips the root folder's own name — selected folders appear as
     * top-level directories and selected files as top-level entries.
     * Same-named top-level entries get a "name (1)" suffix rather than being
     * silently dropped: a multi-select from search results can legitimately
     * pick same-named nodes from different folders.
     */
    fun zipResolvedNodes(targets: List<DriveNode>, out: OutputStream) {
        ZipOutputStream(out).use { zw ->
            val topLevel = HashSet<String>()
            for (root in targets) {
                if (root.type == "file") {
                    val name = uniqueTopLevel(topLevel, sanitizeZipPath(root.name))
                    if (name.isBlank() || root.blobPath.isNullOrBlank()) continue
                    zw.putNextEntry(ZipEntry(name))
                    copyBlobInto(zw, root.blobPath)
                    zw.closeEntry()
                    continue
                }

                val descendants = driveService.collectDescendants(root.id)
                val topName = uniqueTopLevel(topLevel, sanitizeZipPath(root.name))
                if (topName.isBlank()) continue
                val seen = HashSet<String>()
                for (d in descendants) {
                    val rel = if (d.id == root.id) {
                        topName
                    } else {
                        // relPath starts with the root's own name; swap it for
                        // the (possibly suffixed) reserved top-level name.
                        val sub = sanitizeZipPath(d.relPath.removePrefix(root.name).removePrefix("/"))
                        if (sub.isBlank()) continue
                        "$topName/$sub"
                    }
                    val name = if (d.type == "folder") "$rel/" else rel
                    if (!seen.add(name)) continue
                    zw.putNextEntry(ZipEntry(name))
                    if (d.type == "file" && !d.blobPath.isNullOrBlank()) {
                        copyBlobInto(zw, d.blobPath)
                    }
                    zw.closeEntry()
                }
            }
        }
    }

    fun zipNodes(ids: List<Long>, out: OutputStream) = zipResolvedNodes(zipTargets(ids), out)

    /**
     * Returns the subset of ids that are strict descendants of other ids in
     * the same selection (possible when multi-selecting from search results,
     * where ancestors and descendants can appear side by side).
     */
    private fun nestedSelections(ids: List<Long>): Set<Long> {
        if (ids.size < 2) return emptySet()
        val placeholders = ids.joinToString(",") { "?" }
        // Keep as raw SQL: recursive CTE + dynamic IN clause.
        return dsl.resultQuery(
            """
            WITH RECURSIVE descendants(id) AS (
              SELECT n.id FROM drive_nodes n WHERE n.parent_id IN ($placeholders)
              UNION ALL
              SELECT n.id FROM drive_nodes n JOIN descendants d ON n.parent_id = d.id
            )
            SELECT id FROM descendants WHERE id IN ($placeholders)
            """.trimIndent(),
            *(ids + ids).map { it as Any? }.toTypedArray(),
        ).fetch { it.get("id", Long::class.java) }.toSet()
    }

    /** Streams a stored blob into the zip; a missing blob file is skipped silently. */
    private fun copyBlobInto(zw: ZipOutputStream, blobPath: String) {
        val f = File(driveService.blobAbsPath(blobPath))
        if (f.exists()) f.inputStream().use { it.copyTo(zw) }
    }

    companion object {
        private fun sanitizeZipPath(p: String): String =
            p.split('/')
                .map { it.trim() }
                .filter { it.isNotEmpty() && it != "." && it != ".." }
                .joinToString("/")

        /** Reserves a unique top-level entry name, suffixing "stem (1).ext" on collision. */
        fun uniqueTopLevel(seen: MutableSet<String>, name: String): String {
            if (name.isBlank()) return ""
            var cand = name
            var i = 1
            while (!seen.add(cand)) {
                val dot = name.lastIndexOf('.')
                val (stem, ext) = if (dot <= 0) name to "" else name.substring(0, dot) to name.substring(dot)
                cand = "$stem ($i)$ext"
                i++
            }
            return cand
        }
    }
}

File: api-rs/src/service/drive_zip_service.rs
use crate::model::drive::DriveNode;
use crate::service::drive_service::{DriveError, DriveResult, DriveService};
use async_zip::base::write::ZipFileWriter;
use async_zip::{Compression, ZipEntryBuilder};
use std::collections::HashSet;
use std::path::PathBuf;
use std::sync::Arc;
use tokio::io::AsyncWriteExt;
use tokio_util::compat::FuturesAsyncWriteCompatExt;

pub struct DriveZipService {
    drive: Arc<DriveService>,
}

impl DriveZipService {
    pub fn new(drive: Arc<DriveService>) -> Arc<Self> {
        Arc::new(Self { drive })
    }

    /// Streams a zip archive of the folder to the given async writer (e.g. a
    /// duplex pipe whose other end is used as the response body).
    pub async fn zip_folder<W: tokio::io::AsyncWrite + Unpin>(
        &self,
        folder_id: i64,
        writer: W,
    ) -> DriveResult<()> {
        let root = self.drive.find_by_id(folder_id).await?;
        if root.r#type != "folder" || root.deleted_at.is_some() {
            return Err(DriveError::NotFound);
        }
        let descendants = self.drive.collect_descendants(folder_id).await?;

        let writer = tokio_util::compat::TokioAsyncWriteCompatExt::compat_write(writer);
        let mut zw = ZipFileWriter::new(writer);
        let mut seen = std::collections::HashSet::new();
        let root_name = root.name.clone();

        for d in descendants {
            if d.id == folder_id {
                continue;
            }
            let mut rel = d.rel_path.clone();
            if rel.starts_with(&root_name) {
                rel = rel[root_name.len()..].to_string();
            }
            rel = rel.trim_start_matches('/').to_string();
            rel = sanitize_zip_path(&rel);
            if rel.is_empty() {
                continue;
            }
            if !seen.insert(rel.clone()) {
                continue;
            }
            if d.r#type == "folder" {
                let entry =
                    ZipEntryBuilder::new(format!("{}/", rel).into(), Compression::Deflate).build();
                zw.write_entry_whole(entry, &[])
                    .await
                    .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?;
                continue;
            }
            let Some(blob) = d.blob_path else { continue };
            let abs = self.drive.blob_abs_path(&blob);
            let entry = ZipEntryBuilder::new(rel.into(), Compression::Deflate).build();
            let mut ew = zw
                .write_entry_stream(entry)
                .await
                .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?
                .compat_write();
            match tokio::fs::File::open(&abs).await {
                Ok(mut f) => {
                    if let Err(e) = tokio::io::copy(&mut f, &mut ew).await {
                        let _ = ew.shutdown().await;
                        return Err(DriveError::Io(e));
                    }
                }
                Err(_) => {}
            }
            ew.into_inner()
                .close()
                .await
                .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?;
        }
        zw.close()
            .await
            .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?;
        Ok(())
    }

    /// Resolves a multi-select download into its effective root nodes: ids are
    /// deduplicated, ids nested under other selected ids are dropped (their
    /// content arrives via the ancestor), and deleted/missing nodes are
    /// skipped. Returns NotFound when no valid node remains, so the handler
    /// can still send a clean 404 before writing any body.
    pub async fn resolve_zip_targets(&self, ids: &[i64]) -> DriveResult<Vec<DriveNode>> {
        let mut seen = HashSet::with_capacity(ids.len());
        let mut uniq = Vec::with_capacity(ids.len());
        for &id in ids {
            if seen.insert(id) {
                uniq.push(id);
            }
        }
        let nested = self.nested_selections(&uniq).await?;

        let mut targets = Vec::with_capacity(uniq.len());
        for &id in &uniq {
            if nested.contains(&id) {
                continue;
            }
            match self.drive.find_by_id(id).await {
                Ok(n) if n.deleted_at.is_none() => targets.push(n),
                Ok(_) | Err(DriveError::NotFound) => continue,
                Err(e) => return Err(e),
            }
        }
        if targets.is_empty() {
            return Err(DriveError::NotFound);
        }
        Ok(targets)
    }

    /// Streams a zip archive of the given nodes. Unlike zip_folder — which
    /// strips the root folder's own name — selected folders appear as
    /// top-level directories and selected files as top-level entries.
    /// Same-named top-level entries get a "name (1)" suffix rather than being
    /// silently dropped: a multi-select from search results can legitimately
    /// pick same-named nodes from different folders.
    pub async fn zip_nodes<W: tokio::io::AsyncWrite + Unpin>(
        &self,
        targets: &[DriveNode],
        writer: W,
    ) -> DriveResult<()> {
        let writer = tokio_util::compat::TokioAsyncWriteCompatExt::compat_write(writer);
        let mut zw = ZipFileWriter::new(writer);

        let mut top_level: HashSet<String> = HashSet::new();
        for root in targets {
            if root.r#type == "file" {
                let name = unique_top_level(&mut top_level, &sanitize_zip_path(&root.name));
                if name.is_empty() || root.blob_path.is_none() {
                    continue;
                }
                let abs = root
                    .blob_path
                    .as_deref()
                    .map(|b| self.drive.blob_abs_path(b));
                write_file_entry(&mut zw, &name, abs).await?;
                continue;
            }

            let descendants = self.drive.collect_descendants(root.id).await?;
            let top_name = unique_top_level(&mut top_level, &sanitize_zip_path(&root.name));
            if top_name.is_empty() {
                continue;
            }
            let mut seen: HashSet<String> = HashSet::new();
            for d in descendants {
                let rel = if d.id == root.id {
                    top_name.clone()
                } else {
                    // rel_path starts with the root's own name; swap it for the
                    // (possibly suffixed) reserved top-level name.
                    let mut sub = d.rel_path.clone();
                    if sub.starts_with(&root.name) {
                        sub = sub[root.name.len()..].to_string();
                    }
                    let sub = sanitize_zip_path(sub.trim_start_matches('/'));
                    if sub.is_empty() {
                        continue;
                    }
                    format!("{}/{}", top_name, sub)
                };
                if !seen.insert(rel.clone()) {
                    continue;
                }
                if d.r#type == "folder" {
                    let entry =
                        ZipEntryBuilder::new(format!("{}/", rel).into(), Compression::Deflate)
                            .build();
                    zw.write_entry_whole(entry, &[])
                        .await
                        .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?;
                    continue;
                }
                let abs = d.blob_path.as_deref().map(|b| self.drive.blob_abs_path(b));
                write_file_entry(&mut zw, &rel, abs).await?;
            }
        }
        zw.close()
            .await
            .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?;
        Ok(())
    }

    /// Returns the subset of ids that are strict descendants of other ids in
    /// the same selection (possible when multi-selecting from search results,
    /// where ancestors and descendants can appear side by side).
    async fn nested_selections(&self, ids: &[i64]) -> DriveResult<HashSet<i64>> {
        let mut out = HashSet::new();
        if ids.len() < 2 {
            return Ok(out);
        }
        let ids_json = serde_json::to_string(ids).map_err(anyhow::Error::from)?;
        let rows: Vec<(i64,)> = sqlx::query_as(
            r#"
WITH RECURSIVE selected(id) AS (
  SELECT value FROM json_each(?)
),
descendants(id) AS (
  SELECT n.id FROM drive_nodes n WHERE n.parent_id IN (SELECT id FROM selected)
  UNION ALL
  SELECT n.id FROM drive_nodes n JOIN descendants d ON n.parent_id = d.id
)
SELECT id FROM selected WHERE id IN (SELECT id FROM descendants)"#,
        )
        .bind(ids_json)
        .fetch_all(&self.drive.pool)
        .await?;
        for (id,) in rows {
            out.insert(id);
        }
        Ok(out)
    }
}

/// Writes one file entry, streaming the blob when present; a missing blob
/// file is skipped silently (matches zip_folder's behavior).
async fn write_file_entry<W: futures::io::AsyncWrite + Unpin>(
    zw: &mut ZipFileWriter<W>,
    rel: &str,
    abs: Option<PathBuf>,
) -> DriveResult<()> {
    let entry = ZipEntryBuilder::new(rel.into(), Compression::Deflate).build();
    let mut ew = zw
        .write_entry_stream(entry)
        .await
        .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?
        .compat_write();
    if let Some(abs) = abs {
        if let Ok(mut f) = tokio::fs::File::open(&abs).await {
            if let Err(e) = tokio::io::copy(&mut f, &mut ew).await {
                let _ = ew.shutdown().await;
                return Err(DriveError::Io(e));
            }
        }
    }
    ew.into_inner()
        .close()
        .await
        .map_err(|e| DriveError::Other(anyhow::anyhow!(e)))?;
    Ok(())
}

/// Reserves a unique top-level entry name, suffixing "stem (1).ext" style on
/// collision.
fn unique_top_level(seen: &mut HashSet<String>, name: &str) -> String {
    if name.is_empty() {
        return String::new();
    }
    let ext = std::path::Path::new(name)
        .extension()
        .map(|e| format!(".{}", e.to_string_lossy()))
        .unwrap_or_default();
    let stem = name.strip_suffix(&ext).unwrap_or(name);
    let mut cand = name.to_string();
    let mut i = 1;
    while !seen.insert(cand.clone()) {
        cand = format!("{} ({}){}", stem, i, ext);
        i += 1;
    }
    cand
}

fn sanitize_zip_path(p: &str) -> String {
    let mut out = Vec::new();
    for seg in p.split('/') {
        let s = seg.trim();
        if s.is_empty() || s == "." || s == ".." {
            continue;
        }
        out.push(s);
    }
    out.join("/")
}
